### PR TITLE
Xmin rep

### PR DESCRIPTION
**Pull request by serprex**
QRep based xmin replication has a fault: xmin isn't monotonic & has wraparound issues

Wraparound workaround: grab all records with `0 < age(xmin) <= age(last snapshot xmin)` which'll include some records from before wraparound but so it goes

Logic is copied from qrep code

### CODE REVIEW

File: flow/connectors/postgres/qrep.go [hunk 1]
```diff
@@ -596,6 +549,42 @@ func (c *PostgresConnector) SetupQRepMetadataTables(config *protos.QRepConfig) e
...
 }
 
+func (c *PostgresConnector) PullXminRecordStream(
+	config *protos.QRepConfig,
+	currentTxid *int64,
+	partition *protos.QRepPartition,
+	stream *model.QRecordStream,
+) (int, error) {
+	query := config.Query
+	// TODO if initial replication uses xmin with empty partition id, need to do full scan,
```
Comment (serprex) [inline flow/connectors/postgres/qrep.go:559]: TODO

File: flow/activities/flowable.go [hunk 2]
```diff
@@ -845,3 +845,93 @@ func (a *FlowableActivity) CreateTablesFromExisting(ctx context.Context, req *pr
...
+	}
+	defer connectors.CloseConnector(srcConn)
+
+	dstConn, err := connectors.GetQRepSyncConnector(ctx, config.DestinationPeer)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get qrep destination connector: %w", err)
+	}
+	defer connectors.CloseConnector(dstConn)
+
+	log.Info("replicating xmin\n")
```
Comment (iskakaushik) [inline flow/activities/flowable.go:872]: let's log the flowJobName here as well. Should be in the config object.

File: flow/activities/flowable.go [hunk 3]
```diff
@@ -845,3 +845,93 @@ func (a *FlowableActivity) CreateTablesFromExisting(ctx context.Context, req *pr
...
+
+	var stream *model.QRecordStream
+	bufferSize := shared.FetchAndChannelSize
+	var wg sync.WaitGroup
+
+	var goroutineErr error = nil
+
+	stream = model.NewQRecordStream(bufferSize)
```
Comment (iskakaushik) [inline flow/activities/flowable.go:880]: this could just be `stream := `

**Comment by iskakaushik** [comment]
> `0 <= age(xmin) < lastbatchtxid`

@serprex  should'nt this be `0 <= age(xmin) < age(lastbatchtxid)` ?

**Comment by serprex** [comment]
Potential followups:
1. include configurable wait time between scans for small tables
2. be able to have most scans be since last current txid (like original implementation) & only do snapshot scan every few scans

File: flow/workflows/xmin_flow.go [hunk 4]
```diff
@@ -239,7 +238,10 @@ func XminFlowWorkflow(
 		return fmt.Errorf("xmin replication failed: %w", err)
 	}
 
-	state.LastPartition = &protos.QRepPartition{PartitionId: strconv.FormatInt(lastPartition&0xffffffff, 10)}
+	state.LastPartition = &protos.QRepPartition{
+		PartitionId: uuid.New().String(),
```
Comment (iskakaushik) [inline flow/workflows/xmin_flow.go:242]: generating a uuid in a workflow is not repeatable and causes state issues, you have to do it as a side-effect, see https://github.com/PeerDB-io/peerdb/blob/main/flow/workflows/cdc_flow.go#L126 for an example side effect.

Comment (serprex) [reply flow/workflows/xmin_flow.go:242]: Went with reusing q.runUUID which is generated a bit earlier in this code

File: flow/activities/flowable.go [hunk 5]
```diff
@@ -845,3 +845,93 @@ func (a *FlowableActivity) CreateTablesFromExisting(ctx context.Context, req *pr
...
+	}
+	defer connectors.CloseConnector(dstConn)
+
+	log.Info("replicating xmin\n")
+
+	var stream *model.QRecordStream
+	bufferSize := shared.FetchAndChannelSize
+	var wg sync.WaitGroup
+
+	var goroutineErr error = nil
```
[the remark this answers is not in the document]
Comment (serprex) [reply flow/activities/flowable.go]: The code in question seems like it shouldn't progress when an error is caught earlier on

File: flow/workflows/xmin_flow.go (hunk 4 above)
Comment (iskakaushik) [reply flow/workflows/xmin_flow.go]: lol, i just fixed it. feel free to force overwrite.

File: flow/connectors/postgres/qrep_query_executor.go [hunk 6]
```diff
@@ -321,6 +321,51 @@ func (qe *QRepQueryExecutor) ExecuteAndProcessQueryStream(
 		return 0, fmt.Errorf("[pg_query_executor] failed to begin transaction: %w", err)
 	}
 
+	totalRecordsFetched, err := qe.ExecuteAndProcessQueryStreamWithTx(tx, stream, query, args...)
+	return totalRecordsFetched, err
+}
+
+func (qe *QRepQueryExecutor) ExecuteAndProcessQueryStreamGettingCurrentTxid(
```
Comment (iskakaushik) [inline flow/connectors/postgres/qrep_query_executor.go:328]: nit: consider renaming the method

**Pull request merged**